test/realm_test.dart

```diff
@@ -1182,20 +1201,78 @@ Future<void> main([List<String>? args]) async {
       final dog = Dog('Fido', owner: person);
 
       expect(person, person);
-      expect(person, isNot(1)); 
+      expect(person, isNot(1));
       expect(person, isNot(dog));
 
       realm.write(() {
         realm..add(person)..add(dog);
       });
 
       expect(person, person);
-      expect(person, isNot(1)); 
+      expect(person, isNot(1));
       expect(person, isNot(dog));
 
       final read = realm.query<Person>("name == 'Kasper'");
 
       expect(read, [person]);
     });
   });
+  group('Cycle referenced objects tests:', () {
+    test('Realm adding objects graph', () {
+      var studentMichele = Student(1)
+        ..name = "Michele Ernesto"
+        ..yearOfBirth = 2005;
+      var studentLoreta = Student(2)
+        ..name = "Loreta Salvator"
+        ..yearOfBirth = 2006;
+      var studentPeter = Student(3)
+        ..name = "Peter Ivanov"
+        ..yearOfBirth = 2007;
+
+      var school131 = School("JHS 131")..city = "NY";
+      school131.students.addAll([studentMichele, studentLoreta, studentPeter]);
+
+      var school131Branch1 = School("First branch 131A")
+        ..branchOfSchool = school131
+        ..students.addAll([studentMichele, studentLoreta])
+        ..city = "NY Bronx";
+
+      studentMichele.school = school131Branch1;
+      studentLoreta.school = school131Branch1;
+
+      var school131Branch2 = School("Second branch 131B")
+        ..branchOfSchool = school131
+        ..students.add(studentPeter)
+        ..city = "NY Bronx";
+
+      studentPeter.school = school131Branch2;
+
+      school131.branches.addAll([school131Branch1, school131Branch2]);
+
+      var config = Configuration([School.schema, Student.schema]);
+      var realm = Realm(config);
+
+      realm.write(() => realm.add(school131));
+
+      //Check schools
+      var schools = realm.all<School>();
+      expect(schools.length, 3);
+
+      //Check students
+      var students = realm.all<Student>();
+      expect(students.length, 3);
+
+      //Check branches
+      var branches = realm.all<School>().query(r'branchOfSchool != nil');
+      expect(branches.length, 2);
+      expect(branches[0].students.length + branches[1].students.length, 3);
+
+      //Check main schools
+      var mainSchools = realm.all<School>().query(r'branchOfSchool = nil');
```

Comment:
```suggestion
      var mainSchools = realm.all<School>().query('branchOfSchool = nil');
```

Reply:
r"", means regex. and is only useful when we have a query with arguments like query(r"name == $0", [myname]) or one could write query("name == \$0", [myname]) escaping the `$`